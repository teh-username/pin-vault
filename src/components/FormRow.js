import React from 'react';
import { FormLabel, FormInput } from 'react-native-elements';

const FormRow = ({ label, name, value, onInputChange, editable = true }) => [
  <FormLabel key={`${name}_label`}>{label}</FormLabel>,
  <FormInput
    key={`${name}input`}
    value={value}
    onChangeText={data => onInputChange(name, data)}
    placeholder={`Please enter ${label.toLowerCase()}...`}
    editable={editable}
  />,
];

export default FormRow;
