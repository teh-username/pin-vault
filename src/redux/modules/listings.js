import { combineReducers } from 'redux';

export const ADD_ENTRY = 'modules/listings/ADD_ENTRY';
export const MODIFY_ENTRY = 'modules/listings/MODIFY_ENTRY';
export const DELETE_ENTRY = 'modules/listings/DELETE_ENTRY';

/*
  Sample state:
  {
    entries: {
      2cf05d: {
        name: 'Bank 1',
        code: '1234',
      },
      5d94db: {
        name: 'Bank 2',
        code: '5678',
      },
    },
  }
*/

const initialState = {
  entries: {},
};

export const entries = (state = initialState.entries, action) => {
  switch (action.type) {
    case ADD_ENTRY:
      return {
        ...state,
        [action.id]: {
          code: action.code,
          name: action.name,
        },
      };
    case MODIFY_ENTRY:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          code: action.code,
        },
      };
    case DELETE_ENTRY:
      return Object.entries(state)
        .filter(([id, entryDetails]) => action.id !== id)
        .reduce((previousState, [id, entryDetails]) => {
          return {
            ...previousState,
            [id]: entryDetails,
          };
        }, {});
    default:
      return state;
  }
};

export default combineReducers({
  entries,
});

export const addEntry = (id, name, code) => ({
  type: ADD_ENTRY,
  id,
  name,
  code,
});

export const modifyEntry = (id, code) => ({
  type: MODIFY_ENTRY,
  id,
  code,
});

export const deleteEntry = id => ({
  type: DELETE_ENTRY,
  id,
});

export const getEntries = ({ listings: { entries } }) =>
  Object.entries(entries);
export const getEntryDetails = ({ listings: { entries } }, id) => entries[id];
