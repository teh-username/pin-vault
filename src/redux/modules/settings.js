import { combineReducers } from 'redux';

export const TOGGLE_PASSCODE_REQUIREMENT =
  'modules/settings/TOGGLE_PASSCODE_REQUIREMENT';
export const SET_PASSCODE = 'modules/settings/SET_PASSCODE';

/*
  Sample state:
  {
    requirePasscode:  true,
    passcodeHash: '5E884898DA28047151D0E56F8DC6292773603D0D6AABBDD62A11EF721D1542D8',
  }
 */

const initialState = {
  requirePasscode: false,
  passcodeHash: null,
};

const requirePasscode = (state = initialState.requirePasscode, action) => {
  switch (action.type) {
    case TOGGLE_PASSCODE_REQUIREMENT: {
      return !state;
    }
    default:
      return state;
  }
};

const passcodeHash = (state = initialState.passcodeHash, action) => {
  switch (action.type) {
    case SET_PASSCODE:
      return action.passcodeHash;
    default:
      return state;
  }
};

export const togglePasscodeRequirement = () => ({
  type: TOGGLE_PASSCODE_REQUIREMENT,
});

export const setPasscode = passcodeHash => ({
  type: SET_PASSCODE,
  passcodeHash,
});

export default combineReducers({
  requirePasscode,
  passcodeHash,
});

export const getCurrentPasscode = ({ settings: { passcodeHash } }) =>
  passcodeHash;

export const getPasscodeRequirement = ({ settings: { requirePasscode } }) =>
  requirePasscode;
