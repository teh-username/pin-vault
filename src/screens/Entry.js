import React from 'react';
import { connect } from 'react-redux';
import { View, StyleSheet } from 'react-native';
import { Button } from 'react-native-elements';

import FormRow from '../components/FormRow';

import {
  ADD_ENTRY,
  addEntry,
  modifyEntry,
  getEntryDetails,
} from '../redux/modules/listings';
import generateId from '../utils/idGenerator';

const modes = {
  add: 'ENTRY_ADD',
  modify: 'ENTRY_MODIFY',
};

class Entry extends React.Component {
  static navigationOptions = {
    title: 'New Entry',
  };

  constructor(props) {
    super(props);
    const action = props.navigation.state.params.action;
    this.state = {
      id: props.entryDetails ? props.entryDetails.id : null,
      name: props.entryDetails ? props.entryDetails.name : null,
      code: props.entryDetails ? props.entryDetails.code : null,
      mode: action === ADD_ENTRY ? modes.add : modes.modify,
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  handleInputChange(stateKey, data) {
    this.setState({
      [stateKey]: data,
    });
  }

  handleSubmit() {
    const { id, name, code } = this.state;
    this.props.handleSubmit(id, name, code);
    this.props.navigation.goBack();
  }

  render() {
    const { name, code, mode } = this.state;
    return (
      <View>
        <FormRow
          label="Name"
          name="name"
          value={name}
          onInputChange={this.handleInputChange}
          editable={mode === modes.add}
        />
        <FormRow
          label="Code"
          name="code"
          value={code}
          onInputChange={this.handleInputChange}
        />
        <Button
          title={mode === modes.add ? 'Add' : 'Modify'}
          icon={{ name: mode === modes.add ? 'add' : 'mode-edit' }}
          onPress={this.handleSubmit}
          buttonStyle={styles.buttonStyle}
          disabled={!name || !code}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  buttonStyle: {
    top: 40,
    backgroundColor: '#669AE0',
  },
});

const mapStateToDispatch = (
  dispatch,
  { navigation: { state: { params: { action } } } }
) => ({
  handleSubmit(id, name, code) {
    if (action === ADD_ENTRY) {
      dispatch(addEntry(generateId(), name, code));
    } else {
      dispatch(modifyEntry(id, code));
    }
  },
});

const mapStateToProps = (
  state,
  { navigation: { state: { params: { id } } } }
) => ({
  entryDetails: id ? { ...getEntryDetails(state, id), id } : null,
});

export default connect(mapStateToProps, mapStateToDispatch)(Entry);
