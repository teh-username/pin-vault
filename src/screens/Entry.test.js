import React from 'react';
import { Button } from 'react-native-elements';

import { Entry } from './Entry';
import FormRow from '../components/FormRow';
import { ADD_ENTRY, MODIFY_ENTRY } from '../redux/modules/listings';

describe('Entry screen test', () => {
  const navigationMock = {
    state: {
      params: {
        action: ADD_ENTRY,
      },
    },
  };

  const entryDetailsMock = {
    id: 'mockId',
    name: 'mockName',
    code: 'mockCode',
  };

  it('should render as expected', () => {
    const wrapper = shallow(<Entry navigation={navigationMock} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when adding an entry', () => {
    const wrapper = shallow(<Entry navigation={navigationMock} />);
    expect(wrapper.state()).toEqual({
      id: null,
      name: null,
      code: null,
      mode: 'ENTRY_ADD',
    });

    const nameInput = wrapper.find(FormRow).at(0);
    const submitButton = wrapper.find(Button);

    expect(nameInput.prop('editable')).toBe(true);
    expect(submitButton.prop('title')).toBe('Add');
    expect(submitButton.prop('icon')).toEqual({
      name: 'add',
    });
  });

  it('should render correctly when editing an entry', () => {
    const navMock = {
      state: {
        params: {
          action: MODIFY_ENTRY,
        },
      },
    };
    const wrapper = shallow(
      <Entry navigation={navMock} entryDetails={entryDetailsMock} />
    );
    expect(wrapper.state()).toEqual({
      id: 'mockId',
      name: 'mockName',
      code: 'mockCode',
      mode: 'ENTRY_MODIFY',
    });

    const nameInput = wrapper.find(FormRow).at(0);
    const submitButton = wrapper.find(Button);

    expect(nameInput.prop('editable')).toBe(false);
    expect(submitButton.prop('title')).toBe('Modify');
    expect(submitButton.prop('icon')).toEqual({
      name: 'mode-edit',
    });
  });

  it('should disable the button when either or both input is empty', () => {
    const wrapper = shallow(<Entry navigation={navigationMock} />);
    const submitButton = wrapper.find(Button);
    expect(submitButton.prop('disabled')).toBe(true);
  });

  it('should enable the button when both are non-empty', () => {
    const wrapper = shallow(
      <Entry navigation={navigationMock} entryDetails={entryDetailsMock} />
    );
    const submitButton = wrapper.find(Button);
    expect(submitButton.prop('disabled')).toBe(false);
  });

  it('should navigate back on entry submission', () => {
    const navMock = {
      ...navigationMock,
      goBack: jest.fn(),
    };
    const submitMock = jest.fn();
    const wrapper = shallow(
      <Entry
        navigation={navMock}
        entryDetails={entryDetailsMock}
        handleSubmit={submitMock}
      />
    );

    wrapper.instance().handleSubmit();
    expect(navMock.goBack.mock.calls.length).toBe(1);
    expect(submitMock.mock.calls.length).toBe(1);
    expect(submitMock.mock.calls[0]).toEqual([
      'mockId',
      'mockName',
      'mockCode',
    ]);
  });
});
