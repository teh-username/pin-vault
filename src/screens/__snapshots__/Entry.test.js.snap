// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry screen test should render as expected 1`] = `
<View>
  <FormRow
    editable={true}
    label="Name"
    name="name"
    onInputChange={[Function]}
    value={null}
  />
  <FormRow
    label="Code"
    name="code"
    onInputChange={[Function]}
    value={null}
  />
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "#669AE0",
        "top": 40,
      }
    }
    disabled={true}
    icon={
      Object {
        "name": "add",
      }
    }
    onPress={[Function]}
    title="Add"
  />
</View>
`;
